jms/src/main/scala/akka/stream/alpakka/jms/JmsConnector.scala

```diff
@@ -134,6 +134,11 @@ private[jms] class JmsAckSession(override val connection: jms.Connection,
 
   def ack(message: jms.Message): Unit = ackQueue.put(message.acknowledge _)
 
+  override def closeSession(): Unit = {
+    ackQueue.put(() => throw new java.lang.IllegalStateException("Shutdown"))
+    session.close()
+  }
+
```

Comment:
I can't take credit for this I think it might have been you (I just copied it for `shutdown` from `abort`) :)

The mechanism works in `onMessage` by catching this `IllegalStateException` and setting the `listenerStopped` variable to `true`. I can't find any evidence there is anything else throwing `IllegalStateException` so a custom one sounds like a good idea I'll change.